### DISCUSSION TITLE
Redirection to mobile domain in the wild

### DISCUSSION THREAD

**Discussion by ibnesayeed** [cs531-f22/cs531-f22.github.io#18] [Assignment]
Find up to three different "in the wild" redirection examples of the type `example.com` --> `m.example.com` using any mobile user-agent.

**Comment by calebkbrad** [comment]
Below is a curl request to Facebook that uses the -A flag to make the server think the request is coming from a mobile device and also follows any redirects. The first response is a 302 redirect to m.facebook.com (mobile version). 
```
$ curl -IL -A "Mozilla/5.0 (iPhone; CPU iPhone OS 12_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148" https://www.facebook.com
HTTP/2 302
location: https://m.facebook.com/
strict-transport-security: max-age=15552000; preload
content-type: text/html; charset="utf-8"
x-fb-debug: N0dqh3oDjJWWUhRBPE0ZCtxnh1VH82hrTNtyLIzFv7GNFhCrwjITk1H3kEKefluehZTUTBf8QA4x+BsKTrt9RA==
content-length: 0
date: Tue, 18 Oct 2022 02:10:13 GMT
alt-svc: h3=":443"; ma=86400

HTTP/2 200
vary: Accept-Encoding
set-cookie: datr=BQtOYz-x2FHYXzP9mYtczX5f; expires=Thu, 17-Oct-2024 02:10:13 GMT; Max-Age=63072000; path=/; domain=.facebook.com; secure; httponly
set-cookie: fr=08VdLUc8W1wvMKUo0..BjTgsF.3L.AAA.0.0.BjTgsF.AWW2KVdCSJI; expires=Mon, 16-Jan-2023 02:10:12 GMT; Max-Age=7775999; path=/; domain=.facebook.com; secure; httponly
set-cookie: sb=BQtOY1GX6eT-foC_IXL1aiKo; expires=Thu, 17-Oct-2024 02:10:13 GMT; Max-Age=63072000; path=/; domain=.facebook.com; secure; httponly
report-to: {"max_age":259200,"endpoints":[{"url":"https:\/\/m.facebook.com\/ajax\/mtouch_error_reports\/"}]}
x-fb-rlafr: 0
document-policy: force-load-at-top
cross-origin-opener-policy: same-origin-allow-popups
pragma: no-cache
```

Below is much the same, but for youtube.com. It also 302 redirects to the mobile version.
```
$ curl -IL -A "Mozilla/5.0 (iPhone; CPU iPhone OS 12_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148" https://www.youtube.com
HTTP/2 302
content-type: application/binary
x-content-type-options: nosniff
cache-control: no-cache, no-store, max-age=0, must-revalidate
pragma: no-cache
expires: Mon, 01 Jan 1990 00:00:00 GMT
date: Tue, 18 Oct 2022 02:28:30 GMT
location: https://m.youtube.com/
content-length: 0
x-frame-options: SAMEORIGIN
strict-transport-security: max-age=31536000
report-to: {"group":"youtube_main","max_age":2592000,"endpoints":[{"url":"https://csp.withgoogle.com/csp/report-to/youtube_main"}]}
permissions-policy: ch-ua-arch=*, ch-ua-bitness=*, ch-ua-full-version=*, ch-ua-full-version-list=*, ch-ua-model=*, ch-ua-wow64=*, ch-ua-platform=*, ch-ua-platform-version=*
server: ESF
x-xss-protection: 0
alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"

HTTP/2 200
content-type: text/html; charset=utf-8
x-content-type-options: nosniff
cache-control: no-cache, no-store, max-age=0, must-revalidate
pragma: no-cache
expires: Mon, 01 Jan 1990 00:00:00 GMT
date: Tue, 18 Oct 2022 02:28:30 GMT
content-length: 270972
x-frame-options: SAMEORIGIN
strict-transport-security: max-age=31536000
permissions-policy: ch-ua-arch=*, ch-ua-bitness=*, ch-ua-full-version=*, ch-ua-full-version-list=*, ch-ua-model=*, ch-ua-wow64=*, ch-ua-platform=*, ch-ua-platform-version=*
report-to: {"group":"youtube_mobile","max_age":2592000,"endpoints":[{"url":"https://csp.withgoogle.com/csp/report-to/youtube_mobile"}]}
p3p: CP="This is not a P3P policy! See http://support.google.com/accounts/answer/151657?hl=en for more info."
server: ESF
x-xss-protection: 0
set-cookie: GPS=1; Domain=.youtube.com; Expires=Tue, 18-Oct-2022 02:58:31 GMT; Path=/; Secure; HttpOnly
set-cookie: YSC=XyP4Cj0zLKo; Domain=.youtube.com; Path=/; Secure; HttpOnly
set-cookie: VISITOR_INFO1_LIVE=x9TmCFz6qPA; Domain=.youtube.com; Expires=Sun, 16-Apr-2023 02:28:31 GMT; Path=/; Secure; HttpOnly
alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
```

A request to twitch.tv also redirects to a mobile url.
```
$ curl -IL -A "Mozilla/5.0 (iPhone; CPU iPhone OS 12_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148" https://www.twitch.tv/
HTTP/1.1 302 Found
Connection: close
Content-Length: 0
Retry-After: 0
Location: https://m.twitch.tv/?desktop-redirect=true
Date: Tue, 18 Oct 2022 02:49:32 GMT
Set-Cookie: twitch.lohp.countryCode=US; domain=.twitch.tv; expires=Sun, 12 Nov 2023 02:49:32 GMT;
Vary: User-Agent

HTTP/1.1 200 OK
Connection: keep-alive
Content-Type: text/html; charset=utf-8
X-Frame-Options: deny
Content-Security-Policy: frame-ancestors 'none'
Cache-Control: private, no-store
Date: Tue, 18 Oct 2022 02:49:32 GMT
Set-Cookie: unique_id=JzuRmJUejJGnEPjbBYK5ctN3uq0X8vDm; domain=.twitch.tv; path=/; Secure; SameSite=None; expires=Sun, 12 Nov 2023 02:49:32 GMT;
Set-Cookie: unique_id_durable=JzuRmJUejJGnEPjbBYK5ctN3uq0X8vDm; domain=.twitch.tv; path=/; Secure; HttpOnly; SameSite=None; expires=Sun, 12 Nov 2023 02:49:32 GMT;
Set-Cookie: unique_id_samesite_compat=JzuRmJUejJGnEPjbBYK5ctN3uq0X8vDm; domain=.twitch.tv; path=/; Secure; expires=Sun, 12 Nov 2023 02:49:32 GMT;
X-Bucket-Id: e
Vary: Accept-Encoding
```

**Comment by rohitrahiya** [comment]
Hi Professor, I have made a curl request to IMDB.com, snapdeal.com, alibaba.com where the server thinks the request is generated from a mobile device and redirect it to mobile site.

$ curl -IL -A "Mozilla/5.0 (iPhone; CPU iPhone OS 12_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148" https://www.imdb.com/ | grep -iE "^(http|location|Vary)"
% Total % Received % Xferd Average Speed Time Time Time Current
Dload Upload Total Spent Left Speed
0 0 0 0 0 0 0 0 --:--:-- --:--:-- --:--:-- 0
0 801k 0 0 0 0 0 0 --:--:-- --:--:-- --:--:-- 0
HTTP/2 302
location: https://m.imdb.com/
HTTP/2 200
vary: Accept-Encoding,Content-Type,Accept-Encoding,User-Agent

$ curl -IL -A "Mozilla/5.0 (iPhone; CPU iPhone OS 12_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148" https://www.snapdeal.com/ | grep -iE "^(http|location|Vary)"

% Total % Received % Xferd Average Speed Time Time Time Current
Dload Upload Total Spent Left Speed
0 0 0 0 0 0 0 0 --:--:-- --:--:-- --:--:-- 0
0 0 0 0 0 0 0 0 --:--:-- 0:00:01 --:--:-- 0
HTTP/2 302
location: https://m.snapdeal.com/
HTTP/2 200

$ curl -IL -A "Mozilla/5.0 (iPhone; CPU iPhone OS 12_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148" https://www.alibaba.com/ | grep -iE "^(http|location|Vary)"
% Total % Received % Xferd Average Speed Time Time Time Current
Dload Upload Total Spent Left Speed
0 0 0 0 0 0 0 0 --:--:-- --:--:-- --:--:-- 0
0 0 0 0 0 0 0 0 --:--:-- --:--:-- --:--:-- 0
HTTP/2 301
location: https://m.alibaba.com/
HTTP/2 200

**Comment by Badjedi04** [comment]
Hello Professor, Below are curl commands for redirection to mobile version. Response 302 redirects to the mobile version in all three examples(www.shein.com, www.aliexpress.com, www.timesofindia.com)
' ' ' ' 
$ curl -IL -A "Mozilla/5.0 (iPhone; CPU iPhone OS 12_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148" https://www.shein.com 
HTTP/2 302 
date: Thu, 20 Oct 2022 02:35:39 GMT
content-type: text/html
location: https://m.shein.com/us/?ref=www&rep=dir&ret=mus
cf-ray: 75ce5c81ae8f0680-IAD
cf-cache-status: DYNAMIC
set-cookie: _cfuvid=xIFRdxzxK3EThf9lbmRuDox5RfzJYsbZKppjBSxsN1Y-1666233339260-0-604800000; path=/; domain=.shein.com; HttpOnly
server: cloudflare

HTTP/2 200 
date: Thu, 20 Oct 2022 02:35:39 GMT
content-type: text/html; charset=utf-8
cf-ray: 75ce5c83e9f458c6-IAD
cache-control: no-cache, max-age=0, no-store, must-revalidate
set-cookie: pwa_default_currency=USD; expires=Thu, 20 Oct 2022 02:55:39 GMT; max-age=1200; path=/
strict-transport-security: max-age=7776000000; includeSubDomains
cf-cache-status: DYNAMIC
content-security-policy: frame-ancestors https://oa.sheincorp.cn https://activity-admin.manage-aws.sheincorp.cn https://www.messenger.com https://www.facebook.com https://csp.sheincorp.cn https://sqs-admin.biz.sheincorp.cn https://activity-admin.biz.sheincorp.cn https://activity-admin.biz.sheinbackend.com https://sqs-admin.biz.sheinbackend.com https://sqs-admin-gray01.biz.sheinbackend.com https://ccc.biz.sheincorp.cn https://ccc-store.biz.sheincorp.cn https://ccc-store.shein.com https://m.shein.com
nginx-cache: EXPIRED
origin: origin-us.shein.com
referrer-policy: no-referrer-when-downgrade
via-shein-gateway: c-shein-pwa
x-content-type-options: nosniff
x-dns-prefetch-control: on
x-download-options: noopen
x-frame-options: ALLOW-FROM https://www.messenger.com
x-xss-protection: 1; mode=block
set-cookie: cookieId=2031E59F_D478_4674_DB5A_379A316C9F95; expires=Sun, 17 Oct 2032 02:35:39 GMT; max-age=315360000; path=/
set-cookie: cdn_key=pwuslang%3Dus; expires=Thu, 27 Oct 2022 02:35:39 GMT; max-age=604800; path=/
set-cookie: sessionID_shein_m_pwa=s%3AeALvaersygli1Cayz6fnawH_hJqNu_L3.%2FAxpUEXwaHNkbaVMm536QFz6w25Oz4VS1AOgw7xvWw; expires=Fri, 21 Oct 2022 14:35:39 GMT; path=/; HttpOnly; secure
set-cookie: _cfuvid=vXtfQD3gnSpwjuWeuUQBFAKxF43aOchBn9fiLKSmGmg-1666233339664-0-604800000; path=/; domain=.shein.com; HttpOnly
server: cloudflare
' ' ' '

' ' ' '
$ curl -IL -A "Mozilla/5.0 (iPhone; CPU iPhone OS 12_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148" https://www.aliexpress.com
HTTP/2 302 
content-type: text/html
server: Tengine
location: https://m.aliexpress.com/?tracelog=wwwhome2mobilesitehome
access-control-allow-origin: https://hz.aliexpress.com
eagleeye-traceid: 2101d91e16662334197812983e319b
strict-transport-security: max-age=31536000
timing-allow-origin: *
date: Thu, 20 Oct 2022 02:36:59 GMT
set-cookie: ali_apache_id=33.1.217.30.1666233419782.187433.1; path=/; domain=.aliexpress.com; expires=Wed, 30-Nov-2084 01:01:01 GMT
x-akamai-fwd-auth-sha: 4FFF62BE0951C4E230BCE3B8ACAB12429AFD1AAF09F496D35E482559EBA45533
x-akamai-fwd-auth-data: 1164220540, 70.186.29.126, 1666233419, 70.177.211.164
x-akamai-fwd-auth-sign: /9jUQPJFmuKYS/OyRPRY4hCjNdXnuWGXGNvBTMDmOfhGdR2US9+z9siY0/WSSW53NXQncm4EbCMEWCDw0BCrOjoJ9buU1+Afj87IEYDS398=

HTTP/2 302 
server: Tengine
p3p: CP="CAO PSA OUR"
location: https://m.aliexpress.us/?tracelog=wwwhome2mobilesitehome&gatewayAdapt=glo2usa&_randl_shipto=US
eagleeye-traceid: 2101e9d116662334199978174e4968
strict-transport-security: max-age=31536000
timing-allow-origin: *
date: Thu, 20 Oct 2022 02:37:00 GMT
set-cookie: ali_apache_id=33.1.233.209.1666233419998.185024.1; path=/; domain=.aliexpress.com; expires=Wed, 30-Nov-2084 01:01:01 GMT
set-cookie: xman_us_f=x_locale=en_US&x_l=0&x_c_chg=1&acs_rt=d8fd5c1ffe5b41488221d8c7a4999e63; Domain=.aliexpress.com; Expires=Tue, 07-Nov-2090 05:51:07 GMT; Path=/
set-cookie: acs_usuc_t=x_csrf=k74mh_vrrrm_&acs_rt=d8fd5c1ffe5b41488221d8c7a4999e63; Domain=.aliexpress.com; Path=/
set-cookie: intl_locale=en_US; Domain=.aliexpress.com; Path=/
set-cookie: aep_usuc_f=site=usa&c_tp=USD&region=US&b_locale=en_US; Domain=.aliexpress.com; Expires=Tue, 07-Nov-2090 05:51:07 GMT; Path=/
set-cookie: xman_t=/e68kylAnuU85D3YBbRKUwKaAI6pzxSXAz8rPdAxiCXgPWCNyX4EADwGpPIQZNR2; Domain=.aliexpress.com; Expires=Wed, 18-Jan-2023 02:37:00 GMT; Path=/; HttpOnly
set-cookie: intl_common_forever=eobKpK4b69xIFYrnBXWvCgOP9nmM6t42rxA11yhsJIZlqOcVNCR4lw==; Domain=.aliexpress.com; Expires=Tue, 07-Nov-2090 05:51:07 GMT; Path=/; HttpOnly
set-cookie: xman_f=cn7nbSaMDVru7MAt5zsGaXIXuNV2dTSzYko2kVUgtSb/7lDdHru4uR0eu3ktWiDR+kcQ3LAo3+InGRcnTQbg39L3i7Ohdn1+tBJkpXlGoYbicuzsZYyNPw==; Domain=.aliexpress.com; Expires=Tue, 07-Nov-2090 05:51:07 GMT; Path=/; HttpOnly
x-akamai-fwd-auth-sha: 9E301533CBDE77DF360316BD238BA233EBAEC9EF51B46E625FADCE55567DEF2B
x-akamai-fwd-auth-data: 1497878874, 70.186.29.126, 1666233420, 70.177.211.164
x-akamai-fwd-auth-sign: m5em/Kb1fQsrH+DHZhLK0/XqrU9IVNP5HoUIJnDWbQhuhBYbenJEjYqSPqX21M6nLFNMEJ2S86n7Sox5gUfKbpAe4cNZDT/a17Vn7tg4QgQ=
set-cookie: e_id=pt40; Expires=Sun, 17 Oct 2032 02:36:59 GMT; Path=/; Domain=.aliexpress.com

HTTP/2 200 
content-type: text/html;charset=UTF-8
server: Tengine
p3p: CP="CAO PSA OUR"
x-application-context: global-biz-gateway:9901
eagleeye-traceid: 2101e9d516662334203074735ea8e2
strict-transport-security: max-age=31536000
timing-allow-origin: *
date: Thu, 20 Oct 2022 02:37:00 GMT
set-cookie: ali_apache_id=33.1.233.213.1666233420309.192557.6; path=/; domain=.aliexpress.com; expires=Wed, 30-Nov-2084 01:01:01 GMT
set-cookie: account_v=1; Domain=.aliexpress.com; Expires=Tue, 07-Nov-2090 05:51:07 GMT; Path=/
set-cookie: intl_common_forever=2WSGfnWzVzP5hni4Gz7hPLserGgbUMLcQ2jAWYqZehhMMDJR/a6PBA==; Domain=.aliexpress.us; Expires=Tue, 07-Nov-2090 05:51:07 GMT; Path=/; HttpOnly
set-cookie: intl_locale=en_US; Domain=.aliexpress.us; Path=/
set-cookie: xman_us_f=x_locale=en_US&x_l=0&x_c_chg=1; Domain=.aliexpress.us; Expires=Tue, 07-Nov-2090 05:51:07 GMT; Path=/
set-cookie: xman_f=SnPXvqwEFxU68JbfYxpLCghKj6kF0VkrHZZoiQCxhF2PYKEHIMovmlkk0A+YkgFa; Domain=.aliexpress.us; Expires=Tue, 07-Nov-2090 05:51:07 GMT; Path=/; HttpOnly
set-cookie: acs_usuc_t=x_csrf=3ci019iyg23u&acs_rt=6a17c87a3c12411bb84db6d5f16a83d8; Domain=.aliexpress.us; Path=/
set-cookie: xman_t=wqajaN1LpmV3sCmvq/lU1aLLIIGQo1qR8EedoxxBArs+jqR7uR+oAZRJ/CaHCR+4; Domain=.aliexpress.us; Expires=Wed, 18-Jan-2023 02:37:00 GMT; Path=/; HttpOnly
set-cookie: aep_usuc_f=site=usa&c_tp=USD&region=US&b_locale=en_US; Domain=.aliexpress.us; Expires=Tue, 07-Nov-2090 05:51:07 GMT; Path=/
x-akamai-fwd-auth-sha: 7F3B0346D1E233C4002BF591EEC891D4DBB5BDC0150318CD71935A0BD59C2F3D
x-akamai-fwd-auth-data: 1811800843, 70.186.29.126, 1666233420, 70.177.211.164
x-akamai-fwd-auth-sign: M6lNyRL3TiAOE5s5Bzq/FW6b04NXqXE3rpna0BkXs7ohxLOyL7+HRZjy060M7SQa/yc4vyablyUsnP02adCdizaqpljdro7ScBTkHwJUTsg=
set-cookie: e_id=pt60; Expires=Sun, 17 Oct 2032 02:37:00 GMT; Path=/; Domain=.aliexpress.com

' ' ' '

' ' ' '
$ curl -IL -A "Mozilla/5.0 (iPhone; CPU iPhone OS 12_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148" https://www.timesofindia.com
HTTP/2 301 
server: Bhoot
date: Thu, 20 Oct 2022 02:38:46 GMT
content-type: text/html; charset=iso-8859-1
location: https://timesofindia.indiatimes.com/
x-frame-options: sameorigin
strict-transport-security: max-age=25920000; includeSubdomains
x-xss-protection: 1; mode=block
x-content-type-options: nosniff

HTTP/2 301 
server: AkamaiGHost
content-length: 0
location: https://timesofindia.indiatimes.com/us
expires: Thu, 20 Oct 2022 02:38:46 GMT
cache-control: max-age=0, no-cache, no-store
pragma: no-cache
date: Thu, 20 Oct 2022 02:38:46 GMT
access-control-max-age: 86400
access-control-allow-credentials: false
access-control-allow-headers: Origin,X-Requested-With,Content-Type,Accept
access-control-allow-methods: GET,POST
strict-transport-security: max-age=86400

HTTP/2 302 
server: AkamaiGHost
content-length: 0
location: https://m.timesofindia.com/us
expires: Thu, 20 Oct 2022 02:38:46 GMT
cache-control: max-age=0, no-cache, no-store
pragma: no-cache
date: Thu, 20 Oct 2022 02:38:46 GMT
access-control-max-age: 86400
access-control-allow-credentials: false
access-control-allow-headers: Origin,X-Requested-With,Content-Type,Accept
access-control-allow-methods: GET,POST
strict-transport-security: max-age=86400

HTTP/2 200 
x-amz-id-2: F0TDi0BhWoDplvFfnAVV//XcSxWZgEEFbvAJ9DW4DSWKUgJRjdK3jvhamI0t4tLDKwUt1bgQ6yE=
x-amz-request-id: YXSQCWPJCKWD0TDH
last-modified: Thu, 20 Oct 2022 02:37:28 GMT
etag: "c025926d1ece259972e0043f0af2093f"
content-type: text/html; charset=utf-8
server: AmazonS3
expires: Thu, 20 Oct 2022 02:38:46 GMT
cache-control: max-age=0, no-cache, no-store
pragma: no-cache
date: Thu, 20 Oct 2022 02:38:46 GMT
access-control-max-age: 86400
access-control-allow-credentials: false
access-control-allow-headers: Origin,X-Requested-With,Content-Type,Accept
access-control-allow-methods: GET,POST
strict-transport-security: max-age=86400

' ' ' '
